codec/src/test/java/io/netty/handler/codec/MessageAggregatorTest.java

```diff
@@ -0,0 +1,81 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.handler.codec;
+
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufHolder;
+import io.netty.buffer.DefaultByteBufHolder;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.embedded.EmbeddedChannel;
+
+public class MessageAggregatorTest {
+    static class ReadCounter extends ChannelOutboundHandlerAdapter {
+        int value;
+
+        @Override
+        public void read(ChannelHandlerContext ctx) throws Exception {
+            value++;
+            ctx.read();
+        }
+    }
+
+    abstract static class MockMessageAggregator
+        extends MessageAggregator<ByteBufHolder, ByteBufHolder, ByteBufHolder, ByteBufHolder> {
+
+        protected MockMessageAggregator() {
+      super(1024);
+    }
+
+        @Override
+        protected ByteBufHolder beginAggregation(ByteBufHolder start, ByteBuf content) throws Exception {
+            return start.replace(Unpooled.wrappedBuffer(start.content(), content));
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void testReadFlowManagement() throws Exception {
+        ReadCounter counter = new ReadCounter();
+        ByteBufHolder first = new DefaultByteBufHolder(Unpooled.wrappedBuffer("first".getBytes()));
+        ByteBufHolder chunk = new DefaultByteBufHolder(Unpooled.wrappedBuffer("chunk".getBytes()));
+        ByteBufHolder last = new DefaultByteBufHolder(Unpooled.wrappedBuffer("last".getBytes()));
+
+        MockMessageAggregator agg = spy(MockMessageAggregator.class);
+        when(agg.isStartMessage(first)).thenReturn(true);
+        when(agg.isContentMessage(chunk)).thenReturn(true);
+        when(agg.isContentMessage(last)).thenReturn(true);
+        when(agg.isLastContentMessage(last)).thenReturn(true);
+
+        EmbeddedChannel embedded = new EmbeddedChannel(counter, agg);
+        embedded.config().setAutoRead(false);
+
+        assertFalse(embedded.writeInbound(first));
+        assertFalse(embedded.writeInbound(chunk));
+        assertTrue(embedded.writeInbound(last));
+
+        assertEquals(3, counter.value); // 2 reads issued from MockMessageAggregator
+                                                 // 1 read issued from EmbeddedChannel constructor
+        assertTrue(embedded.finish());
+    }
```

Comment:
use `embedded.readInbound()` and so train the read messages. Also assert its what you expect and ensure you release everything